com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/TaskBuilder.java

```diff
@@ -0,0 +1,398 @@
+        return this.executorService.submit(createCallableTask(task, monitor));
+    }
+
+    private boolean compareAllSignatures(byte[] taskSignature, List<IResource> outputResources) {
+        boolean allSigsEquals = true;
+        for (IResource r : outputResources) {
+            byte[] s = state.getSignature(r.getAbsPath());
+            if (!Arrays.equals(s, taskSignature)) {
+                allSigsEquals = false;
+                break;
+            }
+        }
+        return allSigsEquals;
+    }
+
+    // deps are the task input files generated by another task not yet completed,
+    // i.e. "solve" the dependency graph
+    private boolean hasUnresolvedDependencies(Task task) {
+        Set<IResource> dependencies = getUnresolvedDependencies(task);
+        return !dependencies.isEmpty();
+    }
+    private Set<IResource> getUnresolvedDependencies(Task task) {
+        Set<IResource> dependencies = new HashSet<>(task.getInputs());
+        dependencies.retainAll(allOutputs);
+        synchronized (completedOutputs) {
+            dependencies.removeAll(completedOutputs);
+        }
+        return dependencies;
+    }
+
+
+    private boolean checkIfResourcesExist(final List<IResource> resources) {
+        // do all output files exist?
+        boolean allResourcesExists = true;
+        for (IResource r : resources) {
+            if (!r.exists()) {
+                allResourcesExists = false;
+                break;
+            }
+        }
+        return allResourcesExists;
+    }
+
+    private TaskResult buildTask(Task task, IProgress monitor) throws IOException {
+        BundleHelper.throwIfCanceled(monitor);
+
+        logger.info("Build task " + task.getName() + " " + task.getInputsString() + " -> " + task.getOutputsString());
+
+        // if (hasUnresolvedDependencies(task)) {
+        //     // postpone task. dependent input not yet generated
+        //     // logger.warning("NOT ALL DEPS " + task.getInputsString() + " -> " + task.getOutputsString());
+        //     return null;
+        // }
+
+        final List<IResource> outputResources = task.getOutputs();
+
+        boolean allOutputsExist = checkIfResourcesExist(outputResources);
+
+        // compare all task signature. current task signature between previous
+        // signature from state on disk
+        TimeProfiler.start("compare signatures");
+        TimeProfiler.addData("color", "#FFC0CB");
+        TimeProfiler.addData("main input", String.valueOf(task.input(0)));
+        final byte[] taskSignature = task.calculateSignature();
+        boolean allSigsEquals = compareAllSignatures(taskSignature, outputResources);
+        TimeProfiler.stop();
+
+        boolean isCompleted = completedTasks.contains(task);
+        boolean shouldRun = !isCompleted && (!allOutputsExist || !allSigsEquals ||
+                            (buildContainsChanges && task.getBuilder().isGameProjectBuilder()));
```

Comment:
It will not work right without `isLastTask` check

if it's `game.project` and some files were built, but not all of them your code will build game project task, but you shouldn't until very end

Reply:
Hmm, but won't it always build GameProject last?

Reply:
it is, without this extra condition: `(buildContainsChanges && task.getBuilder().isGameProjectBuilder()))`
but we need this extra condition because of https://github.com/defold/defold/pull/9553

Reply:
I think it fails now because of that

Reply:
I've refactored that part of the code now and it should behave correctly